documentation/external_ssl_termination.markdown

```diff
@@ -5,21 +5,61 @@ canonical: "/puppetserver/latest/external_ssl_termination.html"
 
-To allow Puppet Server to recognize the HTTP headers described, open `config.d/master.conf` and add  `allow-header-cert-info: true` to the `webserver` config block. See [Puppet Server Configuration](./configuration.markdown) for more information on the `master.conf` file. Without `allow-header-cert-info` set to 'true', none of these HTTP headers are recognized by Puppet Server.
+## Allow Client Cert Data From HTTP Headers
+
+When using external SSL termination, Puppet Server expects to receive client certificate information via some HTTP headers.
 
-You'll need to reboot Puppet Server for the changes to take effect. The `ssl_client_header` and `ssl_client_verify_header` options in the `puppet.conf` file will then be enabled and work exactly as documented in Ruby Puppet. Please see the [Puppet Documentation](https://docs.puppetlabs.com/references/3.7.latest/configuration.html#sslclientheader) for more detail on these.
+By default, reading this data from headers is disabled. To allow Puppet Server to recognize it, open `config.d/master.conf` and add  `allow-header-cert-info: true` to the `webserver` config block. See [Puppet Server Configuration](./configuration.markdown) for more information on the `master.conf` file.
 
 > **WARNING**: Setting `allow-header-cert-info` to 'true' puts Puppet Server in an incredibly vulnerable state. Take extra caution to ensure it is **absolutely not reachable** by an untrusted network.
+>
+> With `allow-header-cert-info` set to 'true', core Ruby Puppet application code will use only the client HTTP header values---not an SSL-layer client certificate---to determine the client subject name, authentication status, and trusted facts. This is true even if the web server is hosting an HTTPS connection. This applies to validation of the client via rules in the [auth.conf](https://docs.puppetlabs.com/guides/rest_auth_conf.html) file and any [trusted facts](https://docs.puppetlabs.com/puppet/latest/reference/lang_facts_and_builtin_vars.html#trusted-facts) extracted from certificate extensions.
+>
+> If the `client-auth` setting in the `webserver` config block is set to `need` or `want`, the Jetty web server will still validate the client certificate against a certificate authority store, but it will only verify the SSL-layer client certificate---not a certificate in an  `X-Client-Cert` header.
+
+
+## Restart Puppet Server
+
+You'll need to restart Puppet Server for the configuration changes to take effect.
+
+## Configure SSL Terminating Proxy to Set HTTP Headers
+
+The device that terminates SSL for Puppet Server must extract information from the client's certificate and insert that information into three HTTP headers. See the documentation for your SSL terminator for details.
+
+The headers you'll need to set are `X-Client-Verify`, `X-Client-DN`, and `X-Client-Cert`.
+
+### `X-Client-Verify`
+
+Mandatory. Must be one of `NONE`, `SUCCESS`, `GENEROUS` or `FAILED:reason`. Puppet Server uses this to authorize requests.
```

Comment:
Do you know what the specific behaviors of `NONE`, `GENEROUS`, and `FAILED:reason` are and why/when someone would want to use them?  I hadn't seen those before and the documentation at https://docs.puppetlabs.com/references/3.7.latest/configuration.html#sslclientverifyheader only refers to the failures of 'SUCCESS' and *anything else* as "failure".